data/src/main/java/com/microsoft/azure/kusto/data/Utils.java

```diff
@@ -81,12 +79,116 @@ static String post(String url, String payload, InputStream stream, Integer timeo
                     throw new DataServiceException(url, message, ex);
                 }
             }
-        } catch (JSONException | IOException | URISyntaxException e) {
+        } catch (JSONException | IOException e) {
             throw new DataClientException(url, "Error in post request:" + e.getMessage(), e);
```

Comment:
Should the HttpClient and HttpResponse be disposed here?

Reply:
Yes, it ideally should, but we're doing something unique here.

Since we're passing the InputStream externally for the user to consume, I had 2 options:
1. After calling async "httpClient.execute", we could get the CompletableFuture, on which we can call get(). This is a blocking call, and would first read the entire stream from the endpoint, at which point we'd close the httpClient and httpResponse, and only then pass the InputStream. By this point, the InputStream is no longer a true stream from the endpoint. This negates the performance improvement the customer is seeking, as they would like to stream the data as it's received from their ADX cluster.
2. Pass the InputStream as it's streaming from the ADX cluster. If we close the httpClient or the httpResponse in the Utils POST call (before returning the InputStream to the caller and before they consume all its data), reading from the stream will result in an exception that it cannot read from the stream because it's closed. Yet we MUST close those 2 resources. So my solution was to have the resources close AFTER the caller finishes consuming the data from the InputStream. To do this, we should only have to tell the caller they're expected to close the InputStream. Per https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/fundamentals.html closing the stream closes the other resources: "In order to ensure proper release of system resources one must close either the content stream associated with the entity or the response itself.". However, to be extra sure, I wrap the InputStream in an EofSensorInputStream, which closes the stream on its own when all of the data is consumed. Lastly, to be extra sure, I created a new InputStream object to wrap that which includes the HttpClient and HttpResponse objects, and closes those as well explicitly when it's closed.

I included this in my comment at https://github.com/Azure/azure-kusto-java/pull/170/files#diff-9b40b1e0eb4f281e2b999bcc168625e2979a8052a985db6a172c255b0768a1f1R12

Reply:
Sorry, I misunderstood your comment. Yes, we need to close the resources in the Exception.